src/workflow/workflow_tasks/concurrency_manager.rs

```diff
@@ -0,0 +1,284 @@
+use crate::{
+    errors::WorkflowUpdateError,
+    protos::{
+        coresdk::workflow_activation::WfActivation, temporal::api::common::v1::WorkflowExecution,
+    },
+    protosext::ValidPollWFTQResponse,
+    workflow::{
+        workflow_tasks::{OutstandingActivation, OutstandingTask},
+        HistoryUpdate, Result, WorkflowError, WorkflowManager,
+    },
+};
+use futures::future::{BoxFuture, FutureExt};
+use parking_lot::{RwLock, RwLockReadGuard, RwLockWriteGuard};
+use std::{
+    collections::HashMap,
+    fmt::Debug,
+    ops::{Deref, DerefMut},
+};
+use tokio::sync::Mutex;
+
+/// Provides a thread-safe way to access workflow machines for specific workflow runs
+pub(crate) struct WorkflowConcurrencyManager {
+    /// Maps run id -> data about and machines for that run
+    machines: RwLock<HashMap<String, ManagedRun>>,
```

Comment:
This is not `machines`, it's managed runs, should we rename the property?

Reply:
yes